### DISCUSSION TITLE
[Question] about advertise-port and secure-port

### DISCUSSION THREAD

**Discussion by rufherg** [k3s-io/k3s#6108] [Q&A]
Hi, Now I am learning k3s and rancher. But i have a question about the default config for k3s server start.

K3S Version: v1.2.36+k3s1

https://github.com/k3s-io/k3s/blob/v1.23.6+k3s1/pkg/cli/cmds/server.go#L181
https://github.com/k3s-io/k3s/blob/v1.23.6+k3s1/pkg/daemons/control/server.go#L233-L243
As the code above said, the `HTTPSPort` default value is 6443 and the `AdvertisePort ` will equals `HTTPSPort`, and then the `APIServerPort ` value is 6444. That is why the cmd is `--advertise-port=6443 --secure-port=6444` when we start kube-apiserver with k3s.

But after testing, the port of 6443 and 6444 looks like have no difference. They both need to be authorized. So why not to use the same port like 6443 by default?

**Comment by brandond** [comment]
We run the apiserver within k3s, and wrap its HTTP response handler inside our own on port 6443 so that we can expose both the K3s supervisor API and the Kubernetes API on the same port. There is no way to run the apiserver without disabling the built-in listener though, so we had to move it off to another port in order to avoid fatal errors when they both try to use 6443.

**Reply by rufherg** [reply]
Thanks for your answer~ But I can't find some document about K3s supervisor API introduction. So I still don't know what is the difference about K3s superivisor API and the K8s API. Can u give me some web page link to help me know what is K3s supervisor API?

Thanks for your help~^_^
@brandond

**Reply by brandond** [reply]
The K3s supervisor API is used to provision configuration and certificates for new nodes as they join the cluster. It is not really documented anywhere, but you can find the request router at: https://github.com/k3s-io/k3s/blob/1d42f46cda8b7603f43b4c3dcf1e8c13c5c6e014/pkg/server/router.go#L42

The Kubernetes API is everything you would normally see served by the Kubernetes apiserver.

**Reply by rufherg** [reply]
Thank u. But why we can access k8s API on the port 6444? Is it due to the K3s supervisor API forwarding the request to 6443？This is the last question about this discussion.

Thanks for help~
@brandond

**Reply by brandond** [reply]
> But why we can access k8s API on the port 6444? Is it due to the K3s supervisor API forwarding the request to 6443？

Because both APIs are served on that port. Having both the Kubernetes and Supervisor APIs on a single port so that we don't have to expose multiple ports was a design goal for K3s.
